File: thurstone/inference.py

from __future__ import annotations
from dataclasses import dataclass, field
from typing import List, Sequence, Optional, Tuple
import numpy as np
from .density import Density
from .lattice import UniformLattice
from .pricing import Race, StatePricer
from .order_stats import winner_of_many, expected_payoff_with_multiplicity
from .clustering import ClusterSplitter

# ---- Densities from offsets ----

def densities_from_offsets(base: Density, offsets: Sequence[float]) -> List[Density]:
    out = []
    for o in offsets:
        out.append(base.shift_fractional(o))
    return out

def state_prices_from_densities(densities: Sequence[Density]) -> List[float]:
    race = Race(list(densities))
    return list(race.state_prices())

# ---- Implicit map: offset -> price ----

def implicit_state_prices(base: Density, offsets: Sequence[float]) -> List[float]:
    dens = densities_from_offsets(base, offsets)
    return state_prices_from_densities(dens)

# ---- Calibration (inverse) ----

@dataclass
class AbilityCalibrator:
    base: Density
    offset_grid: Optional[Sequence[float]] = None
    n_iter: int = 3
    # Optional per-runner scale handling for 2D calibration (loc, scale)
    scales: Optional[np.ndarray] = field(default=None, repr=False)
    scale_span: float = 0.5          # +/- around each scale value (physical units of scale)
    scale_steps: int = 3             # number of scale samples per runner
    loc_span: float = 5.0            # +/- location range (physical units) for 2D path
    loc_step: float = 0.25           # step for location grid (physical units)
    skew_a: float = 0.0              # skew-normal 'a' used in density_for

    def __post_init__(self):
        if self.offset_grid is None:
            L = self.base.lattice.L
            self.offset_grid = list(range(int(-L/2), int(L/2)))[::-1]

    def set_scales(self, scales: Sequence[float]) -> None:
        self.scales = np.asarray(scales, dtype=float)

    def density_for(self, loc: float, scale: float) -> Density:
        lat = self.base.lattice
        return Density.skew_normal(lat, loc=loc, scale=scale, a=self.skew_a)

    def solve_from_prices(self, prices: Sequence[float], *, initial_offsets: Optional[Sequence[float]]=None) -> List[float]:
        prices_arr = np.asarray(prices, dtype=float)
        n = len(prices_arr)
        # 2D path if scales provided and length matches; else 1D fallback
        if self.scales is None or len(self.scales) != n:
            if initial_offsets is None:
                initial_offsets = [0.0 for _ in prices_arr]
            unit = self.base.lattice.unit
            # Work internally in lattice steps
            offsets = np.array(initial_offsets, dtype=float) / unit

            for _ in range(self.n_iter):
                new_offsets = offsets.copy()
                grid = list(self.offset_grid)
                # Update each runner independently via monotone interpolation,
                # conditioning on current offsets of others.
                for i in range(n):
                    # Tabulate the implicit curve p_i(g) with others fixed
                    implied_prices_i = []
                    cur = offsets.copy()
                    for g in grid:
                        cur[i] = g
                        pvec = implicit_state_prices(self.base, cur)
                        implied_prices_i.append(pvec[i])
                    implied_prices_i = np.array(implied_prices_i, dtype=float)
                    # Enforce monotone xp by sorting and uniquing keys
                    pairs = sorted(zip(implied_prices_i, grid), key=lambda t: t[0])
                    xp = np.array([t[0] for t in pairs], dtype=float)  # prices
                    fp = np.array([t[1] for t in pairs], dtype=float)  # offsets
                    xp_unique, idx = np.unique(xp, return_index=True)
                    fp_unique = fp[idx]
                    target = float(prices_arr[i])
                    target = float(np.clip(target, xp_unique.min(), xp_unique.max()))
                    gi = float(np.interp(target, xp_unique, fp_unique))
                    new_offsets[i] = gi
                offsets = new_offsets

            # Convert lattice-step offsets back to ability units
            return list(np.array(offsets, dtype=float) * unit)

        # ---- 2D calibration in (loc, scale) per runner ----
        scales = np.asarray(self.scales, dtype=float)
        unit = self.base.lattice.unit
        # Initialize locs (abilities) in physical units
        if initial_offsets is None:
            locs = np.zeros(n, dtype=float)
        else:
            locs = np.asarray(initial_offsets, dtype=float) * unit

        # Location grid (physical units) around 0
        loc_grid = np.arange(-self.loc_span, self.loc_span + self.loc_step, self.loc_step, dtype=float)

        def scale_grid_for(si: float) -> np.ndarray:
            if self.scale_steps <= 1:
                return np.array([max(1e-6, si)], dtype=float)
            offsets_scale = np.linspace(-self.scale_span, self.scale_span, self.scale_steps, dtype=float)
            sg = si + offsets_scale
            return np.maximum(1e-6, sg.astype(float))

        for _ in range(self.n_iter):
            # Precompute field distribution once per iteration (current locs/scales), multiplicity-aware
            current_densities = [self.density_for(float(locs[j]), float(scales[j])) for j in range(n)]
            densityAll, multAll = winner_of_many(current_densities)
            cdfAll = densityAll.cdf()
            for i in range(n):
                si = float(scales[i])
                pi_target = float(prices_arr[i])
                sg = scale_grid_for(si)
                loc_estimates = []
                for s in sg:
                    # Build p_i(loc | scale=s) curve with others fixed
                    p_curve = []
                    for loc_candidate in loc_grid:
                        d_i_candidate = self.density_for(float(loc_candidate), float(s))
                        payoff_vec = expected_payoff_with_multiplicity(d_i_candidate, densityAll, multAll, cdf=None, cdfAll=cdfAll)
                        p_i = float(np.sum(payoff_vec))
                        p_curve.append(p_i)
                    p_curve = np.array(p_curve, dtype=float)
                    pairs = sorted(zip(p_curve, loc_grid), key=lambda t: t[0])
                    xp = np.array([pp for pp, _ in pairs], dtype=float)
                    fp = np.array([ll for _, ll in pairs], dtype=float)
                    p_clamped = float(np.clip(pi_target, xp.min(), xp.max()))
                    loc_s = float(np.interp(p_clamped, xp, fp))
                    loc_estimates.append(loc_s)
                loc_estimates = np.array(loc_estimates, dtype=float)
                if len(sg) == 1:
                    locs[i] = loc_estimates[0]
                else:
                    locs[i] = float(np.interp(si, sg, loc_estimates))

        # Remove translation (identifiability) by median-centering
        locs = locs - np.median(locs)
        return list(locs)

    def solve_from_dividends(self, dividends: Sequence[float], nan_value: float=2000.0) -> List[float]:
        prices = StatePricer.prices_from_dividends(dividends, nan_value=nan_value)
        return self.solve_from_prices(prices)

    # forward direction (ability -> state prices/dividends), with clustering for extended offsets
    def state_prices_from_ability(self, ability: Sequence[float]) -> List[float]:
        offsets = [a / self.base.lattice.unit for a in ability]
        splitter = ClusterSplitter()
        return splitter.extended_state_prices(self.base, offsets)

    def dividends_from_ability(self, ability: Sequence[float], multiplicity: float=1.0) -> List[float]:
        prices = self.state_prices_from_ability(ability)
        return list(1.0/(multiplicity*np.array(prices)))

File: docs/js/thurstone/inference.js
// Port of thurstone/inference.py
// AbilityCalibrator.solve_from_dividends / solve_from_prices via per-iteration global curve.
// The 2D (loc, scale) path is included; use set_scales() before calling solve.

import { Density } from "./density.js";
import { StatePricer } from "./pricing.js";
import { winnerOfMany, expectedPayoffWithMultiplicity } from "./order_stats.js";
import { ClusterSplitter } from "./clustering.js";

export function densitiesFromOffsets(base, offsets) {
  return offsets.map(o => base.shiftFractional(o));
}

// numpy.interp equivalent. xp must be strictly ascending.
export function interp(x, xp, fp) {
  if (xp.length === 0) return 0;
  if (x <= xp[0]) return fp[0];
  if (x >= xp[xp.length - 1]) return fp[fp.length - 1];
  // binary search
  let lo = 0, hi = xp.length - 1;
  while (hi - lo > 1) {
    const mid = (lo + hi) >>> 1;
    if (xp[mid] <= x) lo = mid; else hi = mid;
  }
  const t = (x - xp[lo]) / (xp[hi] - xp[lo]);
  return fp[lo] + t * (fp[hi] - fp[lo]);
}

export function sortAscByFirst(pairs) {
  pairs.sort((a, b) => a[0] - b[0]);
  return pairs;
}

export function uniqueByFirst(xp, fp) {
  // np.unique semantics: keep first occurrence
  const outX = [], outF = [];
  let prev = NaN;
  for (let i = 0; i < xp.length; i++) {
    if (xp[i] !== prev) {
      outX.push(xp[i]);
      outF.push(fp[i]);
      prev = xp[i];
    }
  }
  return [Float64Array.from(outX), Float64Array.from(outF)];
}

export function median(arr) {
  const a = [...arr].sort((x, y) => x - y);
  const n = a.length;
  if (n === 0) return 0;
  return n % 2 ? a[(n - 1) >>> 1] : 0.5 * (a[n / 2 - 1] + a[n / 2]);
}

export class AbilityCalibrator {
  constructor(base, {
    offsetGrid = null,
    nIter = 3,
    scales = null,
    scaleSpan = 0.5,
    scaleSteps = 3,
    locSpan = 5.0,
    locStep = 0.25,
    skewA = 0.0,
  } = {}) {
    this.base = base;
    this.nIter = nIter;
    this.scales = scales == null ? null : Float64Array.from(scales);
    this.scaleSpan = scaleSpan;
    this.scaleSteps = scaleSteps;
    this.locSpan = locSpan;
    this.locStep = locStep;
    this.skewA = skewA;
    this.lookupCurve1dPrices = null;
    this.lookupCurve1dInverse = null;
    this.lookupCurves2dPrices = new Map();
    this.lookupCurves2dInverse = new Map();

    if (offsetGrid == null) {
      const L = this.base.lattice.L;
      const lo = -Math.trunc(L / 2);
      const hi = Math.trunc(L / 2);
      const grid = [];
      for (let g = lo; g < hi; g++) grid.push(g);
      grid.reverse();    // match Python [::-1] on range(int(-L/2), int(L/2))
      this.offsetGrid = grid;
    } else {
      this.offsetGrid = [...offsetGrid];
    }
  }

  setScales(scales) { this.scales = Float64Array.from(scales); }

  densityFor(loc, scale) {
    return Density.skewNormal(this.base.lattice, { loc, scale, a: this.skewA });
  }

  rebuildCurves1d(locs) {
    const unit = this.base.lattice.unit;
    const offsets = locs.map(l => l / unit);
    const grid = this.offsetGrid;
    const densField = densitiesFromOffsets(this.base, offsets);
    const { density: dAll, mult: multAll } = winnerOfMany(densField);
    const cdfAll = dAll.cdf();
    const implied = new Float64Array(grid.length);
    for (let i = 0; i < grid.length; i++) {
      const dg = this.base.shiftFractional(grid[i]);
      const ep = expectedPayoffWithMultiplicity(dg, dAll, multAll, null, cdfAll);
      let s = 0; for (let k = 0; k < ep.length; k++) s += ep[k];
      implied[i] = s;
    }
    // Cache loc -> price curve (sorted ascending in loc)
    const locsPhys = grid.map(g => unit * g);
    const ord = locsPhys.map((_, k) => k).sort((a, b) => locsPhys[a] - locsPhys[b]);
    const locsSorted = ord.map(i => locsPhys[i]);
    const pricesSorted = ord.map(i => implied[i]);
    this.lookupCurve1dPrices = { locs: Float64Array.from(locsSorted), prices: Float64Array.from(pricesSorted) };
    // Cache inverse curve (sorted ascending in price, unique)
    const pairs = grid.map((g, i) => [implied[i], g]);
    sortAscByFirst(pairs);
    const xp = pairs.map(p => p[0]);
    const fp = pairs.map(p => p[1]);
    const [xu, fu] = uniqueByFirst(xp, fp);
    this.lookupCurve1dInverse = { prices: xu, offsets: fu };
  }

  rebuildCurves2d(locs, scales) {
    const n = locs.length;
    const densField = [];
    for (let j = 0; j < n; j++) densField.push(this.densityFor(locs[j], scales[j]));
    const { density: dAll, mult: multAll } = winnerOfMany(densField);
    const cdfAll = dAll.cdf();
    const locGrid = [];
    for (let v = -this.locSpan; v <= this.locSpan + 1e-9; v += this.locStep) locGrid.push(+v);
    const uniqueScales = [...new Set(scales.map(s => +s))].sort((a, b) => a - b);
    this.lookupCurves2dPrices.clear();
    this.lookupCurves2dInverse.clear();
    for (const s of uniqueScales) {
      const pCurve = new Float64Array(locGrid.length);
      for (let i = 0; i < locGrid.length; i++) {
        const d = this.densityFor(locGrid[i], s);
        const ep = expectedPayoffWithMultiplicity(d, dAll, multAll, null, cdfAll);
        let sum = 0; for (let k = 0; k < ep.length; k++) sum += ep[k];
        pCurve[i] = sum;
      }
      this.lookupCurves2dPrices.set(s, { locs: Float64Array.from(locGrid), prices: pCurve });
      const pairs = locGrid.map((l, i) => [pCurve[i], l]);
      sortAscByFirst(pairs);
      const [xu, fu] = uniqueByFirst(pairs.map(p => p[0]), pairs.map(p => p[1]));
      this.lookupCurves2dInverse.set(s, { prices: xu, offsets: fu });
    }
  }

  solveFromPrices(prices, { initialOffsets = null } = {}) {
    const pricesArr = Float64Array.from(prices);
    const n = pricesArr.length;

    if (this.scales == null || this.scales.length !== n) {
      // 1D path
      const unit = this.base.lattice.unit;
      let offsets = (initialOffsets == null ? Array(n).fill(0) : [...initialOffsets]).map(o => o / unit);
      const grid = this.offsetGrid;
      for (let iter = 0; iter < this.nIter; iter++) {
        const densField = densitiesFromOffsets(this.base, offsets);
        const { density: dAll, mult: multAll } = winnerOfMany(densField);
        const cdfAll = dAll.cdf();
        const implied = new Float64Array(grid.length);
        for (let i = 0; i < grid.length; i++) {
          const dg = this.base.shiftFractional(grid[i]);
          const ep = expectedPayoffWithMultiplicity(dg, dAll, multAll, null, cdfAll);
          let s = 0; for (let k = 0; k < ep.length; k++) s += ep[k];
          implied[i] = s;
        }
        // Cache (loc -> price) and (price -> offset)
        const locsPhys = grid.map(g => unit * g);
        const ord = locsPhys.map((_, k) => k).sort((a, b) => locsPhys[a] - locsPhys[b]);
        this.lookupCurve1dPrices = {
          locs: Float64Array.from(ord.map(i => locsPhys[i])),
          prices: Float64Array.from(ord.map(i => implied[i])),
        };
        const pairs = grid.map((g, i) => [implied[i], g]);
        sortAscByFirst(pairs);
        const [xu, fu] = uniqueByFirst(pairs.map(p => p[0]), pairs.map(p => p[1]));
        this.lookupCurve1dInverse = { prices: xu, offsets: fu };
        // Invert all prices at once
        const lo = xu[0], hi = xu[xu.length - 1];
        for (let i = 0; i < n; i++) {
          let p = pricesArr[i];
          if (p < lo) p = lo; if (p > hi) p = hi;
          offsets[i] = interp(p, xu, fu);
        }
      }
      return offsets.map(o => o * unit);
    }

    // 2D (loc, scale) path
    const scales = this.scales;
    const unit = this.base.lattice.unit;
    let locs = initialOffsets == null ? new Float64Array(n) : Float64Array.from(initialOffsets.map(o => o * unit));
    const locGrid = [];
    for (let v = -this.locSpan; v <= this.locSpan + 1e-9; v += this.locStep) locGrid.push(+v);

    const scaleGridFor = (si) => {
      if (this.scaleSteps <= 1) return [Math.max(1e-6, si)];
      const out = new Float64Array(this.scaleSteps);
      const step = (2 * this.scaleSpan) / (this.scaleSteps - 1);
      for (let k = 0; k < this.scaleSteps; k++) out[k] = Math.max(1e-6, si + (-this.scaleSpan + k * step));
      return out;
    };

    for (let iter = 0; iter < this.nIter; iter++) {
      const densField = [];
      for (let j = 0; j < n; j++) densField.push(this.densityFor(locs[j], scales[j]));
      const { density: dAll, mult: multAll } = winnerOfMany(densField);
      const cdfAll = dAll.cdf();
      const perRunnerSg = []; const allS = new Set();
      for (let i = 0; i < n; i++) {
        const sg = scaleGridFor(scales[i]);
        perRunnerSg.push(sg);
        for (const s of sg) allS.add(+s);
      }
      const cache = new Map();
      for (const s of allS) {
        const pCurve = new Float64Array(locGrid.length);
        for (let k = 0; k < locGrid.length; k++) {
          const d = this.densityFor(locGrid[k], s);
          const ep = expectedPayoffWithMultiplicity(d, dAll, multAll, null, cdfAll);
          let sum = 0; for (let j = 0; j < ep.length; j++) sum += ep[j];
          pCurve[k] = sum;
        }
        this.lookupCurves2dPrices.set(s, { locs: Float64Array.from(locGrid), prices: pCurve });
        const pairs = locGrid.map((l, i) => [pCurve[i], l]);
        sortAscByFirst(pairs);
        const [xu, fu] = uniqueByFirst(pairs.map(p => p[0]), pairs.map(p => p[1]));
        cache.set(s, { xu, fu });
        this.lookupCurves2dInverse.set(s, { prices: xu, offsets: fu });
      }
      for (let i = 0; i < n; i++) {
        const si = scales[i], pi = pricesArr[i], sg = perRunnerSg[i];
        const locEsts = new Float64Array(sg.length);
        for (let k = 0; k < sg.length; k++) {
          const { xu, fu } = cache.get(+sg[k]);
          let p = pi; if (p < xu[0]) p = xu[0]; if (p > xu[xu.length - 1]) p = xu[xu.length - 1];
          locEsts[k] = interp(p, xu, fu);
        }
        locs[i] = sg.length === 1 ? locEsts[0] : interp(si, Float64Array.from(sg), locEsts);
      }
    }
    const m = median(Array.from(locs));
    return Array.from(locs, x => x - m);
  }

  solveFromDividends(dividends, { nanValue } = {}) {
    const prices = StatePricer.pricesFromDividends(dividends, nanValue);
    return this.solveFromPrices(prices);
  }

  statePricesFromAbility(ability) {
    const offsets = ability.map(a => a / this.base.lattice.unit);
    return new ClusterSplitter().extendedStatePrices(this.base, offsets);
  }

  dividendsFromAbility(ability, multiplicity = 1.0) {
    const p = this.statePricesFromAbility(ability);
    return p.map(pi => 1.0 / (multiplicity * pi));
  }
}
